ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,411 @@
+/*
+ * Use this function when the number of arrays in n2 is 1.
+ * In this case, there is no need to add back.
+ */
+static inline void bc_fast_div(BC_VECTOR *n1_vector, size_t n1_arr_size, BC_VECTOR n2_vector, BC_VECTOR *quot_vector, size_t quot_arr_size)
 {
-	size_t carry, value;
-	unsigned char *nptr, *rptr;
+	size_t n1_top_index = n1_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+	for (size_t i = 0; i < quot_arr_size - 1; i++) {
+		if (n1_vector[n1_top_index - i] < n2_vector) {
+			quot_vector[quot_top_index - i] = 0;
+			n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+			continue;
+		}
+		quot_vector[quot_top_index - i] = n1_vector[n1_top_index - i] / n2_vector;
+		n1_vector[n1_top_index - i] -= n2_vector * quot_vector[quot_top_index - i];
+		n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+		n1_vector[n1_top_index - i] = 0;
+	}
+	/* last */
+	quot_vector[0] = n1_vector[0] / n2_vector;
+}
 
-	if (digit == 0) {
-		memset(result, 0, size);
-	} else {
-		if (digit == 1) {
-			memcpy(result, num, size);
-		} else {
-			/* Initialize */
-			nptr = (unsigned char *) (num + size - 1);
-			rptr = (unsigned char *) (result + size - 1);
-			carry = 0;
-
-			while (size-- > 0) {
-				value = *nptr-- * digit + carry;
-				*rptr-- = value % BASE;
-				carry = value / BASE;
-			}
+/*
+ * Used when the number of elements in the n2 vector is 2 or more.
+ */
+static inline void bc_standard_div(
+	BC_VECTOR *n1_vector, size_t n1_arr_size,
+	BC_VECTOR *n2_vector, size_t n2_arr_size, size_t n2_len,
+	BC_VECTOR *quot_vector, size_t quot_arr_size
+) {
+	size_t i, j;
+	size_t n1_top_index = n1_arr_size - 1;
+	size_t n2_top_index = n2_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+
+	BC_VECTOR div_carry = 0;
+
+	/*
+	 * If calculate the temporary quotient using only one array of n1 and n2, the error will be greater
+	 * than 1 and the number of additions and backs will increase significantly.
+	 * Therefore, when calculating the quotient, adjust the number of digits for n2 so that it becomes
+	 * BC_VECTOR_SIZE + 1 digit, and adjust the number of digits for n1 by the same amount.
+	 * By doing so, the error in the quotient will be within 1, so if need to add back, only need to add it once.
+	 *
+	 * Details:
+	 * Suppose that when calculating the temporary quotient, only the high-order elements of the BC_VECTOR array are used.
+	 * At this time, n1 and n2 can be considered to be decomposed as follows. (Here, B = 10^b)
+	 * n1 = n1_high * B^k + n1_low
+	 * n2 = n2_high * B^k + n2_low
+	 *
+	 * The error E can be expressed by the following formula.
+	 * E = n1_high * B^k / n2_high * B^k - (n1_high * B^k + n1_low) / (n2_high * B^k + n2_low)
+	 * E = n1_high / n2_high - (n1_high * B^k + n1_low) / (n2_high * B^k + n2_low)
+	 * E = (n1_high * (n2_high * B^k + n2_low) - (n1_high * B^k + n1_low) * n2_high) / n2_high * (n2_high * B^k + n2_low)
```

Comment:
From line 97 to line 98 there is a mistake I think. Isn't the last part supposed to be ` / (n2_high * B^k + n2_low)` instead of ` * (n2_high * B^k + n2_low)` ?

Reply:
Consequently the steps that follow on this could be wrong too.

Reply:
ditto